src/java.base/share/classes/java/lang/Class.java

```diff
@@ -3021,6 +3021,45 @@ private void checkPackageAccess(SecurityManager sm, final ClassLoader ccl,
         }
     }
 
+    /*
+     * Checks if a client loaded in ClassLoader ccl is allowed to access the provided
+     * classes under the current package access policy. If access is denied,
+     * throw a SecurityException.
+     *
+     * NOTE: this method should only be called if a SecurityManager is active
+     *       classes must be non-empty
+     *       all classes provided must be loaded by the same ClassLoader
+     */
+    private static void checkPackageAccessForClasses(SecurityManager sm, final ClassLoader ccl,
+                                    boolean checkProxyInterfaces, Class<?>[] classes) {
+        final ClassLoader cl = classes[0].getClassLoader0();
+
+        if (ReflectUtil.needsPackageAccessCheck(ccl, cl)) {
+            Set<String> packages = new HashSet<>();
+
+            for (Class<?> c : classes) {
+                // skip the package access check on a proxy class in default proxy package
+                if (!Proxy.isProxyClass(c) || ReflectUtil.isNonPublicProxyClass(c)) {
```

Comment:
If a sealed class is in a named module, the permitted subclasses/subinterfaces are in the same module as the sealed class.  If a sealed class is in an unnamed module, it will be in the same runtime package as the sealed class.   A proxy class is dynamically generated and not intended for statically named in `permits` clause of a sealed class`.  It can be in a different module or different package.  So a permitted subclass or interface should never be a proxy class. 

So the package access check for permitted subclasses/subinterfaces can be simplified.  I would suggest it to be inlined in `getPermittedSubclasses` as follows:

```
      SecurityManager sm = System.getSecurityManager();
        if (subclasses.length > 0 && sm != null) {
            ClassLoader ccl = ClassLoader.getClassLoader(Reflection.getCallerClass());
            ClassLoader cl = getClassLoader0();
            if (ReflectUtil.needsPackageAccessCheck(ccl, cl)) {
                Set<String> packages = new HashSet<>();
                for (Class<?> c : subclasses) {
                    if (Proxy.isProxyClass(c))
                        throw new InternalError("a permitted subclass should not be a proxy class: " + c);
                    String pkg = c.getPackageName();
                    if (!pkg.isEmpty())
                        packages.add(pkg);
                }
                for (String pkg : packages) {
                    sm.checkPackageAccess(pkg);
                }
            }
        }
```